src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500RelativeDistinguishedName.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Formats.Asn1;
+
+namespace System.Security.Cryptography.X509Certificates
+{
+    /// <summary>
+    ///   Represents a Relative Distinguished Name component of an X.500 Distinguished Name.
+    /// </summary>
+    /// <seealso cref="X500DistinguishedName"/>
+    public sealed class X500RelativeDistinguishedName
+    {
+        private Oid? _singleElementType;
+        private readonly ReadOnlyMemory<byte> _singleElementValue;
+
+        /// <summary>
+        ///   Gets the encoded representation of this Relative Distinguished Name.
+        /// </summary>
+        /// <value>
+        ///   The encoded representation of this Relative Distinguished Name.
+        /// </value>
+        public ReadOnlyMemory<byte> RawData { get; }
+
+        internal X500RelativeDistinguishedName(ReadOnlyMemory<byte> rawData)
+        {
+            RawData = rawData;
+
+            ReadOnlySpan<byte> rawDataSpan = rawData.Span;
+            AsnValueReader outer = new AsnValueReader(rawDataSpan, AsnEncodingRules.DER);
+
+            // Windows does not enforce the sort order on multi-value RDNs.
+            AsnValueReader rdn = outer.ReadSetOf(skipSortOrderValidation: true);
+            AsnValueReader typeAndValue = rdn.ReadSequence();
+
+            Oid firstType = Oids.GetSharedOrNewOid(ref typeAndValue);
+            ReadOnlySpan<byte> firstValue = typeAndValue.ReadEncodedValue();
+            typeAndValue.ThrowIfNotEmpty();
+
+            if (rdn.HasData)
+            {
+                while (rdn.HasData)
+                {
+                    typeAndValue = rdn.ReadSequence();
+
+                    if (Oids.GetSharedOrNullOid(ref typeAndValue) is null)
+                    {
+                        typeAndValue.ReadObjectIdentifier();
+                    }
+
+                    typeAndValue.ReadEncodedValue();
+                    typeAndValue.ThrowIfNotEmpty();
+                }
+            }
+            else
+            {
+                _singleElementType = firstType;
+
+                if (!rawDataSpan.Overlaps(firstValue, out int offset))
+                {
+                    Debug.Fail("AsnValueReader.ReadEncodedValue returned data outside its bounds");
+                    throw new UnreachableException();
+                }
```

Comment:
This can't just be an assert?

If we actually think it's possible, does the setting of _singleElementType need to move below this so that it's only set if the exception doesn't happen?

Reply:
The full call can't be asserted, since the offset is needed (and isn't ever 0).

```C#
bool overlaps = rawDataSpan.Overlaps(firstValue, out int offset);
Debug.Assert(overlaps, "AsnValueReader.ReadEncodedValue always returns a slice of the input");
```

would be possible... but feels... dirty.

Reply:
Why does it feel dirty?  Personally I find it cleaner than:
```C#
Debug.Fail("This can't ever happen...");
throw new UnreachableException(); // ...but we need an exception to satisfy control flow
```

Reply:
You also don't even need the assert if you don't want it.  If the method were actually called "GetOffset" instead of Overlaps and returned the offset, would you still throw?

Reply:
Well, I'd assume that GetOffset would throw if it didn't make sense... so... indirectly, yes :)

Reply:
The only way we reach this else block is if there's a bug in a component we own, yes?  But you'd be ok with such a bug going unchecked in GetOffset?